File: backend-flutter/src/models/Supplier.js
const mongoose = require('mongoose');

const supplierSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Supplier name is required'],
    trim: true,
    maxlength: [100, 'Supplier name cannot exceed 100 characters']
  },
  contactPerson: {
    name: {
      type: String,
      trim: true,
      maxlength: [100, 'Contact person name cannot exceed 100 characters']
    },
    email: {
      type: String,
      trim: true,
      lowercase: true,
      match: [
        /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/,
        'Please enter a valid email address'
      ]
    },
    phone: {
      type: String,
      trim: true,
      match: [
        /^\+?[\d\s\-\(\)]+$/,
        'Please enter a valid phone number'
      ]
    }
  },
  address: {
    street: {
      type: String,
      trim: true,
      maxlength: [200, 'Street address cannot exceed 200 characters']
    },
    city: {
      type: String,
      trim: true,
      maxlength: [100, 'City cannot exceed 100 characters']
    },
    state: {
      type: String,
      trim: true,
      maxlength: [100, 'State cannot exceed 100 characters']
    },
    zipCode: {
      type: String,
      trim: true,
      maxlength: [20, 'Zip code cannot exceed 20 characters']
    },
    country: {
      type: String,
      trim: true,
      maxlength: [100, 'Country cannot exceed 100 characters']
    }
  },
  businessInfo: {
    website: {
      type: String,
      trim: true,
      match: [
        /^https?:\/\/.+/,
        'Please enter a valid website URL'
      ]
    },
    taxId: {
      type: String,
      trim: true,
      maxlength: [50, 'Tax ID cannot exceed 50 characters']
    },
    paymentTerms: {
      type: String,
      trim: true,
      maxlength: [100, 'Payment terms cannot exceed 100 characters']
    }
  },
  isActive: {
    type: Boolean,
    default: true
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  productCount: {
    type: Number,
    default: 0
  }
}, {
  timestamps: true
});

// Index for better query performance
supplierSchema.index({ name: 1 });
supplierSchema.index({ isActive: 1 });
supplierSchema.index({ createdBy: 1 });
supplierSchema.index({ 'contactPerson.email': 1 });

module.exports = mongoose.model('Supplier', supplierSchema);

File: flutter_app/lib/core/models/supplier.dart
class Supplier {
  final String id;
  final String name;
  final ContactPerson? contactPerson;
  final Address? address;
  final BusinessInfo? businessInfo;
  final bool isActive;
  final String createdBy;
  final int productCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  Supplier({
    required this.id,
    required this.name,
    this.contactPerson,
    this.address,
    this.businessInfo,
    required this.isActive,
    required this.createdBy,
    required this.productCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Supplier.fromJson(Map<String, dynamic> json) {
    return Supplier(
      id: json['_id'] ?? json['id'],
      name: json['name'] ?? '',
      contactPerson: json['contactPerson'] != null 
          ? ContactPerson.fromJson(json['contactPerson'])
          : null,
      address: json['address'] != null 
          ? Address.fromJson(json['address'])
          : null,
      businessInfo: json['businessInfo'] != null 
          ? BusinessInfo.fromJson(json['businessInfo'])
          : null,
      isActive: json['isActive'] ?? true,
      createdBy: json['createdBy']?['firstName'] != null 
          ? '${json['createdBy']['firstName']} ${json['createdBy']['lastName']}'
          : json['createdBy'] ?? '',
      productCount: json['productCount'] ?? 0,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'contactPerson': contactPerson?.toJson(),
      'address': address?.toJson(),
      'businessInfo': businessInfo?.toJson(),
      'isActive': isActive,
      'createdBy': createdBy,
      'productCount': productCount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  Supplier copyWith({
    String? id,
    String? name,
    ContactPerson? contactPerson,
    Address? address,
    BusinessInfo? businessInfo,
    bool? isActive,
    String? createdBy,
    int? productCount,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Supplier(
      id: id ?? this.id,
      name: name ?? this.name,
      contactPerson: contactPerson ?? this.contactPerson,
      address: address ?? this.address,
      businessInfo: businessInfo ?? this.businessInfo,
      isActive: isActive ?? this.isActive,
      createdBy: createdBy ?? this.createdBy,
      productCount: productCount ?? this.productCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

class ContactPerson {
  final String? name;
  final String? email;
  final String? phone;

  ContactPerson({
    this.name,
    this.email,
    this.phone,
  });

  factory ContactPerson.fromJson(Map<String, dynamic> json) {
    return ContactPerson(
      name: json['name'],
      email: json['email'],
      phone: json['phone'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'email': email,
      'phone': phone,
    };
  }
}

class Address {
  final String? street;
  final String? city;
  final String? state;
  final String? zipCode;
  final String? country;

  Address({
    this.street,
    this.city,
    this.state,
    this.zipCode,
    this.country,
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    return Address(
      street: json['street'],
      city: json['city'],
      state: json['state'],
      zipCode: json['zipCode'],
      country: json['country'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'street': street,
      'city': city,
      'state': state,
      'zipCode': zipCode,
      'country': country,
    };
  }
}

class BusinessInfo {
  final String? website;
  final String? taxId;
  final String? paymentTerms;

  BusinessInfo({
    this.website,
    this.taxId,
    this.paymentTerms,
  });

  factory BusinessInfo.fromJson(Map<String, dynamic> json) {
    return BusinessInfo(
      website: json['website'],
      taxId: json['taxId'],
      paymentTerms: json['paymentTerms'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'website': website,
      'taxId': taxId,
      'paymentTerms': paymentTerms,
    };
  }
}
